nwastdlib/asyncio_cache.py

```diff
@@ -0,0 +1,169 @@
+    h = hmac.new(secret.encode(), message, hashlib.sha512)
+    return h.hexdigest()
+
+
+async def set_signed_cache_value(pool: AIORedis, secret: str, cache_key: str, value: Any, expiry: int):
+    pickled_value = pickle.dumps(value)
+    checksum = get_hmac_checksum(secret, pickled_value)
+    pipeline = pool.pipeline()
+    pipeline.setex(cache_key, expiry, pickle.dumps(value)).setex(f"{cache_key}-checksum", expiry, checksum.encode())
+    result1, result2 = await pipeline.execute()
+    if not result1 or not result2:
+        logger.warning("Cache not set", cache_key=cache_key, value_ok=result1, checksum_ok=result2)
+
+
+async def get_signed_cache_value(pool: AIORedis, secret: str, cache_key: str):
+    pipeline = pool.pipeline()
+    pipeline.get(cache_key).get(f"{cache_key}-checksum")
+    pickled_value, cache_checksum = await pipeline.execute()
+    if not pickled_value or not cache_checksum:
+        return None
+
+    cache_checksum = cache_checksum.decode("utf-8")
+    if (checksum := get_hmac_checksum(secret, pickled_value)) != cache_checksum:
+        logger.error(
+            "Checksum for cache was wrong, someone tampered with the values!",
+            correct_checksum=str(cache_checksum),
+            recalculated_checksum=checksum,
+        )
+        return None
+
+    return pickle.loads(pickled_value)
+
+
+def cached_result(pool: AIORedis, prefix: str, secret: str, key_name: str | None, expiry: int = 120) -> Callable:
+    """Pass returned result objects from a function call into redis.
+
+    Returns a decorator function that will cache every result of a function to redis. This only works
+    for functions with string, int or UUID arguments and result objects that can be serialized by the
+    python pickle library.
+
+    Example:
+        Todo: switch to cPickle?
+        Decorate a suitable function with all the default settings like this:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM")
+            def my_cached_function(uuid_arg, string_arg, int_kwarg=1):
+                return do_stuff(uuid_arg, string_arg, kwarg)
+
+        The first call is cached and reused for 120s. If the defaults are inadequate use:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM", "desired_cache_key_name", 120)
+            def my_other_function...
+
+    Args:
+        pool: A async redis cache pool.
+        secret: used to generate a checksum too ensure cache tampering is not possible.
+        prefix: Prefix for the cache keys generated.
+        key_name: When specified the total redis key_name will be "prefix:key_name".
+        expiry: expiration in seconds. Defaults to two minutes (120s).
+
+    Returns:
+        decorator function
+
+    """
+
+    def cache_decorator(func: Callable) -> Callable:
+
+        # Todo: write a sync implementation?
+
+        @wraps(func)
+        async def func_wrapper(*args, **kwargs):
```

Comment:
For mypy, if you define this typevar `R = TypeVar("R")` then it should be possible to type this as 

```
    def cache_decorator(func: Awaitable[R]) -> Awaitable[R]:

        # Todo: write a sync implementation?

        @wraps(func)
        async def func_wrapper(*args, **kwargs) -> R:
```

Reply:
Mypy doesn't seem to like the Awaitable. Had to revert it.